src/app/machines/views/MachineList/MachineListControls/HiddenColumnsSelect/HiddenColumnsSelect.tsx

```diff
@@ -1,44 +1,74 @@
 import { CheckboxInput, ContextualMenu } from "@canonical/react-components";
 
+import type { MachineListControlsProps } from "../MachineListControls";
+
 import { useSendAnalytics } from "app/base/hooks";
-import { columnLabels, columns } from "app/machines/constants";
+import { columnLabels, columnToggles } from "app/machines/constants";
 
 type Props = {
   hiddenColumns: string[];
-  toggleHiddenColumn: (column: string) => void;
-};
+} & Pick<MachineListControlsProps, "setHiddenColumns">;
 
 const HiddenColumnsSelect = ({
   hiddenColumns,
-  toggleHiddenColumn,
+  setHiddenColumns,
 }: Props): JSX.Element => {
   const sendAnalytics = useSendAnalytics();
-
+  const selectedColumnsLength = columnToggles.length - hiddenColumns.length;
+  const someColumnsChecked =
+    selectedColumnsLength > 0 && selectedColumnsLength < columnToggles.length;
+  const toggleHiddenColumn = (column: string): void => {
+    if (hiddenColumns.includes(column)) {
+      setHiddenColumns(hiddenColumns.filter((c) => c !== column));
+    } else {
+      setHiddenColumns([...hiddenColumns, column]);
+    }
+  };
   return (
     <ContextualMenu
       className="filter-accordion"
       constrainPanelWidth
-      dropdownProps={{ "aria-label": "hidden columns menu" }}
+      dropdownProps={{ "aria-label": "columns menu" }}
       hasToggleIcon
       position="left"
       toggleClassName="filter-accordion__toggle"
-      toggleLabel={`Hidden columns ${
-        hiddenColumns.length > 0 ? `(${hiddenColumns.length})` : ""
-      }`}
+      /* TODO: add <Icon name="settings" /> after updating to latest react-components  */
```

Comment:
https://github.com/canonical/maas-ui/pull/4589 <-- Let's make a point of getting this PR fixed so that we can do this

Reply:
I'm working on this, it's nearly done.